embassy/src/util/mpsc.rs

```diff
@@ -0,0 +1,905 @@
+        if self.sender.is_closed() {
+            Poll::Ready(())
+        } else {
+            self.sender
+                .channel
+                .get()
+                .set_senders_waker(cx.waker().clone());
+            Poll::Pending
+        }
+    }
+}
+
+impl<'ch, M, T, const N: usize> Drop for Sender<'ch, M, T, N>
+where
+    M: Mutex<Data = ()>,
+{
+    fn drop(&mut self) {
+        self.channel.get().deregister_sender()
+    }
+}
+
+impl<'ch, M, T, const N: usize> Clone for Sender<'ch, M, T, N>
+where
+    M: Mutex<Data = ()>,
+{
+    #[allow(clippy::clone_double_ref)]
+    fn clone(&self) -> Self {
+        self.channel.get().register_sender();
+        Sender {
+            channel: self.channel.clone(),
+        }
+    }
+}
+
+/// An error returned from the [`try_recv`] method.
+///
+/// [`try_recv`]: super::Receiver::try_recv
+#[derive(PartialEq, Eq, Clone, Copy, Debug)]
+pub enum TryRecvError {
+    /// A message could not be received because the channel is empty.
+    Empty,
+
+    /// The message could not be received because the channel is empty and closed.
+    Closed,
+}
+
+/// Error returned by the `Sender`.
+#[derive(Debug)]
+pub struct SendError<T>(pub T);
+
+impl<T> fmt::Display for SendError<T> {
+    fn fmt(&self, fmt: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(fmt, "channel closed")
+    }
+}
+
+/// This enumeration is the list of the possible error outcomes for the
+/// [try_send](super::Sender::try_send) method.
+#[derive(Debug)]
+pub enum TrySendError<T> {
+    /// The data could not be sent on the channel because the channel is
+    /// currently full and sending would require blocking.
+    Full(T),
+
+    /// The receive half of the channel was explicitly closed or has been
+    /// dropped.
+    Closed(T),
+}
+
+impl<T> fmt::Display for TrySendError<T> {
+    fn fmt(&self, fmt: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(
+            fmt,
+            "{}",
+            match self {
+                TrySendError::Full(..) => "no available capacity",
+                TrySendError::Closed(..) => "channel closed",
+            }
+        )
+    }
+}
+
+struct ChannelState<T, const N: usize> {
+    buf: [MaybeUninit<UnsafeCell<T>>; N],
+    read_pos: usize,
+    write_pos: usize,
+    full: bool,
+    closing: bool,
+    closed: bool,
+    receiver_registered: bool,
+    senders_registered: u32,
+    receiver_waker: Option<Waker>,
+    senders_waker: Option<Waker>,
+}
+
+impl<T, const N: usize> ChannelState<T, N> {
+    const INIT: MaybeUninit<UnsafeCell<T>> = MaybeUninit::uninit();
+
+    const fn new() -> Self {
+        let buf = [Self::INIT; N];
+        let read_pos = 0;
+        let write_pos = 0;
+        let full = false;
+        let closing = false;
+        let closed = false;
+        let receiver_registered = false;
+        let senders_registered = 0;
+        let receiver_waker = None;
+        let senders_waker = None;
+        ChannelState {
+            buf,
```

Comment:
it's nicer to write all the initial values inline: `ChannelState { buf: [xx], read_pos: 0, write_pos: 0...}`

Reply:
Done as per commit 7acf14d9ffbc11216e7a504ec9b7f52639e8f660